# Using the metrics suite, on beams2d

You have trained models and want to know what their designs are worth. This notebook
runs the whole evaluation surface on a real EngiBench problem, with published
checkpoints, the physics metrics included:

| You want to... | You call |
|---|---|
| see every metric and what it measures | `METRICS.explain()` |
| score published checkpoints by name | `Board.load("beams2d", ["vqgan", ...])` |
| score checkpoints and your own constructions under one spec | `Board.from_evaluator(evaluator, models, designs=...)` |
| read the result | `board.explain()`, `board.rank("fog")` |
| ask the same questions in another space | `Board(...).evaluate(board.designs, space="pca")` |
| add a metric of your own | `@register_metric(...)` on a function |

It downloads four checkpoints and the beams2d dataset from the Hub, samples twelve
designs per model, and runs the optimizer from each of them. Expect a few minutes on a
laptop.

## 1. What metrics exist, and what does each one measure?

In [1]:
import warnings

import pandas as pd

from engiopt.evaluation.registry import METRICS

# Importing the metrics package registers the built-in metrics.
import engiopt.evaluation.metrics  # noqa: F401  # isort: skip

warnings.filterwarnings("ignore")
pd.set_option("display.max_colwidth", None)
pd.set_option("display.width", 200)

METRICS.explain()

/opt/anaconda3/envs/EngiBench312/lib/python3.12/site-packages/gymnasium/spaces/box.py:235: UserWarning: WARN: Box low's precision lowered by casting to float32, current low.dtype=float64
  gym.logger.warn(
/opt/anaconda3/envs/EngiBench312/lib/python3.12/site-packages/gymnasium/spaces/box.py:305: UserWarning: WARN: Box high's precision lowered by casting to float32, current high.dtype=float64
  gym.logger.warn(


,question,family,direction,cost
mmd,"Distance between the generated designs and the reference optima, compared as whole sets. Zero means the two sets are indistinguishable.",distribution,lower is better,cheap
dpp,"Spread of the generated set, as the geometric mean of its similarity-kernel eigenvalues. One is perfectly spread, near zero is collapsed.",diversity,higher is better,cheap
train_distance,Per-element distance from each generated design to the nearest design in the training set. Zero means the model reproduces its training data.,memorization,diagnostic,cheap
cond_sens,"How much each design changes, per element, when the model is re-run under another sample's conditions. Zero means the conditions are ignored.",conditions,diagnostic,cheap
viol,Fraction of the generated designs that break a constraint of the problem. Zero means every design is admissible.,feasibility,lower is better,cheap
iog,"Optimality gap of each generated design as generated, before any re-optimization. Zero means already optimal.",performance,lower is better,expensive
cog,"Optimality gap summed over every optimizer step when re-optimizing from each generated design, the area under its gap curve. Zero means no optimizer effort was wasted.",performance,lower is better,expensive
fog,Optimality gap of each design once re-optimization from it has finished. Zero means the warm start reached the reference optimum.,performance,lower is better,expensive
per_condition_distance,Per-element distance from each generated design to the reference optimum for its own conditions. Zero means every design matches its own optimum.,conditions,lower is better,cheap
volume_error,Gap between each design's material fraction and the fraction its conditions requested. Zero means every design hit its volume target.,conditions,lower is better,cheap


Each row names the quantity, says how it is computed, and says what the two ends of
its scale mean. The direction column marks some metrics **diagnostic**: those columns
tell you whether the other columns can be trusted, and are never ranked on.

## 2. The problem, and the evaluation spec

Every published number on beams2d is computed under one frozen spec. The spec fixes
which fifty test conditions are scored, the reference optimum for each, and how
per-design values collapse to one number. The kernel bandwidth behind `mmd`, `vendi`
and `dpp` is the median pairwise distance of the training designs unless the spec pins
a value, and the resolved number is recorded in every row. Two
models scored under the same spec are comparable; two scored under different specs
are not, and every row records which spec produced it.

The dataset has one optimal design per set of conditions. That is what makes a
lookup table a serious competitor here, and it is why the spec draws its conditions
from the withheld test split.

This notebook departs from the published spec in two declared ways, so that it runs in
minutes. It scores twelve conditions rather than fifty, which cuts sampling and
optimizer time by the same factor, and so it drops the published spec's condition
digest, which pins the fifty. And it aggregates per-design values with the median rather
than the mean, because one diverged design drags a mean by orders of magnitude and the
physics columns are read as medians throughout EngiOpt. Numbers here are therefore not
comparable with the public leaderboard; the calls are identical.

In [2]:
import dataclasses

from engiopt.evaluation.evaluator import Evaluator
from engiopt.evaluation.spec import EvalSpec

published = EvalSpec.load("beams2d/v2")
spec = dataclasses.replace(published, n_samples=12, condition_digest=None, aggregation="median")
evaluator = Evaluator.for_problem("beams2d", spec=spec)

REFERENCE = evaluator.resolved.ref_designs  # the withheld optimum for each scored condition
print("conditions scored:", len(REFERENCE), "| design shape:", REFERENCE.shape[1:])
print("condition keys:", evaluator.resolved.condition_keys)
print("kernel bandwidth:", spec.sigma or "median distance of the training designs", "| aggregation:", spec.aggregation)

conditions scored: 12 | design shape: (50, 100)
condition keys: ('volfrac', 'rmin', 'forcedist', 'overhang_constraint')
kernel bandwidth: median distance of the training designs | aggregation: median


## 3. Three constructions built from the dataset

Beside the trained models, three rows whose right score is known in advance. They are
not competitors. They are the probes that show what each column can and cannot see.

| Row | What it does | What it should reveal |
|---|---|---|
| `lookup_table` | for each scored condition, the training design whose conditions are nearest | a model that never generalizes, yet answers each condition plausibly |
| `unconditional` | training designs drawn at random, ignoring the conditions | a model that ignores its conditions entirely |
| `permuted_conditions` | the reference optima, shuffled across the conditions | a perfect set of designs, every one for the wrong conditions |

In [3]:
import numpy as np

train = evaluator.problem.dataset["train"]
TRAIN = np.stack([np.asarray(design) for design in train["optimal_design"]])
train_conditions = np.column_stack([np.asarray(train[key], dtype=float) for key in evaluator.resolved.condition_keys])
scored_conditions = evaluator.resolved.conditions_tensor.cpu().numpy().astype(float)

# Nearest neighbour in condition space, each condition standardized so none dominates by its units.
scale = train_conditions.std(axis=0)
scale[scale == 0] = 1.0
nearest = np.argmin(np.linalg.norm((train_conditions[None] - scored_conditions[:, None]) / scale, axis=2), axis=1)

rng = np.random.default_rng(0)
CONSTRUCTIONS = {
    "lookup_table": TRAIN[nearest],
    "unconditional": TRAIN[rng.integers(0, len(TRAIN), len(REFERENCE))],
    "permuted_conditions": REFERENCE[rng.permutation(len(REFERENCE))],
}
print("training designs:", TRAIN.shape)

training designs: (3880, 50, 100)


## 4. Score the trained models and the constructions under one spec

Four published checkpoints at training seed one. Two of them were trained with a
configuration that is not the script's default, so they are named by the fingerprint
of that configuration rather than by name alone. Everything else is one call.

The physics columns run the optimizer from each design, so each cell below takes a
minute or two. The GAN and VQGAN family first, then the diffusion model, which samples
slowly, then the constructions.

In [4]:
from engiopt.evaluation.board import Board
from engiopt.utils.all_generators import BUILTIN_GENERATORS

EXPENSIVE = True  # set False to skip the optimizer and get the cheap columns in seconds


def load(name, fingerprint=None):
    """The published beams2d checkpoint of a model family at training seed one."""
    return BUILTIN_GENERATORS[name].from_pretrained(
        evaluator.problem, problem_id="beams2d", seed=1, config_fingerprint=fingerprint
    )


fast_models = {
    "cgan_cnn_2d": load("cgan_cnn_2d", "825831f6"),
    "gan_cnn_2d": load("gan_cnn_2d", "6293adb3"),
    "vqgan": load("vqgan"),
}
board = Board.from_evaluator(evaluator, fast_models, expensive=EXPENSIVE)
board.frame.round(4)

Fetching 8 files:   0%|          | 0/8 [00:00<?, ?it/s]

Fetching 8 files:   0%|          | 0/8 [00:00<?, ?it/s]

Fetching 6 files:   0%|          | 0/6 [00:00<?, ?it/s]

,mmd,coverage,vendi,dpp,viol,volume_error,per_condition_distance,cond_sens,train_distance,train_distance_ratio,...,train_minutes,iog,cog,fog,calls_to_near_optimum,gap_after_1_calls,gap_after_2_calls,gap_after_5_calls,gap_after_10_calls,reaches_reference_rate
cgan_cnn_2d,0.5337,0.50,2.0504,0.0266,0.9167,0.0722,0.4144,0.2407,0.2902,14.2969,...,NaN,1.816083e+09,1.735349e+09,3.0758,78.0,1.735347e+09,1018.6727,174.6169,56.5027,0.1667
gan_cnn_2d,0.8228,0.25,1.0290,0.0002,1.0000,0.2407,0.4447,0.0000,0.2581,12.7131,...,NaN,8.950722e+09,8.689641e+09,0.0019,17.5,8.689632e+09,4121.7064,189.3875,47.1883,0.4167
vqgan,0.0096,1.00,5.8642,0.4346,0.3333,0.0058,0.0833,0.4704,0.0353,1.7389,...,NaN,2.413100e+00,8.195000e-01,-0.6194,1.5,2.423100e+00,5.6037,0.5194,0.0135,0.6667


In [5]:
diffusion = Board.from_evaluator(evaluator, {"diffusion_2d_cond": load("diffusion_2d_cond")}, expensive=EXPENSIVE)
board.frame = pd.concat([board.frame, diffusion.frame])
board.designs.update(diffusion.designs)
diffusion.frame.round(4)

,mmd,coverage,vendi,dpp,viol,volume_error,per_condition_distance,cond_sens,train_distance,train_distance_ratio,...,train_minutes,iog,cog,fog,calls_to_near_optimum,gap_after_1_calls,gap_after_2_calls,gap_after_5_calls,gap_after_10_calls,reaches_reference_rate
diffusion_2d_cond,0.1041,1.0,6.0826,0.4784,0.8333,0.1271,0.1532,0.3969,0.143,7.0439,...,NaN,-1.7211,6.0217,-0.9128,4.0,-1.7127,26.3979,6.1904,0.3109,0.9167


In [6]:
probes = Board.from_evaluator(evaluator, {}, designs=CONSTRUCTIONS, expensive=EXPENSIVE)
board.frame = pd.concat([board.frame, probes.frame])
board.designs.update(probes.designs)
board.frame.round(4)

,mmd,coverage,vendi,dpp,viol,volume_error,per_condition_distance,cond_sens,train_distance,train_distance_ratio,...,train_minutes,iog,cog,fog,calls_to_near_optimum,gap_after_1_calls,gap_after_2_calls,gap_after_5_calls,gap_after_10_calls,reaches_reference_rate
cgan_cnn_2d,0.5337,0.50,2.0504,0.0266,0.9167,0.0722,0.4144,0.2407,0.2902,14.2969,...,NaN,1.816083e+09,1.735349e+09,3.0758,78.0,1.735347e+09,1018.6727,174.6169,56.5027,0.1667
gan_cnn_2d,0.8228,0.25,1.0290,0.0002,1.0000,0.2407,0.4447,0.0000,0.2581,12.7131,...,NaN,8.950722e+09,8.689641e+09,0.0019,17.5,8.689632e+09,4121.7064,189.3875,47.1883,0.4167
vqgan,0.0096,1.00,5.8642,0.4346,0.3333,0.0058,0.0833,0.4704,0.0353,1.7389,...,NaN,2.413100e+00,8.195000e-01,-0.6194,1.5,2.423100e+00,5.6037,0.5194,0.0135,0.6667
diffusion_2d_cond,0.1041,1.00,6.0826,0.4784,0.8333,0.1271,0.1532,0.3969,0.1430,7.0439,...,NaN,-1.721100e+00,6.021700e+00,-0.9128,4.0,-1.712700e+00,26.3979,6.1904,0.3109,0.9167
lookup_table,0.0395,1.00,5.3611,0.1796,0.7500,0.0125,0.0569,NaN,0.0000,0.0000,...,NaN,9.574600e+00,7.314000e-01,-0.7363,3.5,9.590800e+00,8.2825,3.3313,0.3676,0.5833
unconditional,0.0902,1.00,7.0133,0.5779,0.9167,0.0562,0.3993,NaN,0.0000,0.0000,...,NaN,4.805870e+03,2.817597e+04,1.4287,101.0,4.806746e+03,1994.0963,66.6704,4.2396,0.1667
permuted_conditions,0.0000,1.00,6.1870,0.2039,0.9167,0.0562,0.4231,NaN,0.0203,1.0000,...,NaN,8.783402e+07,8.788590e+07,4.4569,101.0,8.782727e+07,1865.3040,85.9982,27.0510,0.0000


## 5. Read the board

In [7]:
board.explain()

,question,direction,picks,real designs score
mmd,"Distance between the generated designs and the reference optima, compared as whole sets. Zero means the two sets are indistinguishable.",lower is better,permuted_conditions,None
coverage,Fraction of the reference optima that have at least one generated design close to them. One means every reference design is reached.,higher is better,"tie: vqgan, diffusion_2d_cond, lookup_table, unconditional, permuted_conditions",None
vendi,"Effective number of distinct designs in the generated set. One means all identical, the sample size means all mutually distinct.",higher is better,unconditional,None
dpp,"Spread of the generated set, as the geometric mean of its similarity-kernel eigenvalues. One is perfectly spread, near zero is collapsed.",higher is better,unconditional,None
viol,Fraction of the generated designs that break a constraint of the problem. Zero means every design is admissible.,lower is better,vqgan,None
volume_error,Gap between each design's material fraction and the fraction its conditions requested. Zero means every design hit its volume target.,lower is better,vqgan,None
per_condition_distance,Per-element distance from each generated design to the reference optimum for its own conditions. Zero means every design matches its own optimum.,lower is better,lookup_table,None
cond_sens,"How much each design changes, per element, when the model is re-run under another sample's conditions. Zero means the conditions are ignored.",diagnostic,,None
train_distance,Per-element distance from each generated design to the nearest design in the training set. Zero means the model reproduces its training data.,diagnostic,,None
train_distance_ratio,Per-element distance from each generated design to the nearest design in the training set. Zero means the model reproduces its training data.,diagnostic,,None


Why these seven rows: four trained models and three constructions whose right
answer is known, so the board shows which columns can tell a good model from a
construction that games one metric. How: every row was scored under one spec, twelve
withheld conditions, medians over designs, and the physics columns re-optimized from
each design. What the numbers say:

**`permuted_conditions` posts an MMD of exactly zero and full coverage. It is the
reference set.** It also posts the worst final gap on the board, 4.46, and reaches the
reference optimum from none of its designs. A perfect set of designs, each for the
wrong conditions, is the best model by the distribution columns and the worst by the
physics. `per_condition_distance` reads 0.42 for it, in the same tier as the GANs, and
that is the column that sees the problem before any optimizer runs.

**`lookup_table` never generalizes, every design is a training design, and it beats
every trained model except diffusion on final gap.** It has the best cumulative gap on
the board and the smallest per-condition distance, 0.057. That is not a bug. With one
optimum per condition and a dense training sweep, the nearest training design is a very
good warm start. `train_distance` reads exactly zero for it, and so does the ratio: that
is the column that says what it is. `permuted_conditions` reads a ratio of exactly one,
by construction, because it is the reference set and one is what real unseen designs
read. The trained models read 1.7 for the VQGAN, 7 for diffusion and 13 to 14 for the
GANs: further from the training data than real designs are, which is what blurred or
noisy output looks like. Unlike the data is not the same as generalizing.

**`diffusion_2d_cond` has the best final gap, minus 0.91, and reaches the reference
optimum from eleven designs in twelve.** It also violates a constraint in ten of twelve
designs and misses the requested material fraction by the widest margin of the trained
models, 0.127. Its initial gap is negative, meaning better than the reference optimum
before any optimization, which is what a design that spends more material than it was
allowed looks like. Read `iog` beside `viol` and `volume_error`, never alone.

**`vqgan` is the most obedient model.** Fewest violations, smallest volume error, and
the largest response when its conditions change. Its physics is middling. Obedience
and optimality are different columns.

**`cgan_cnn_2d` and `gan_cnn_2d` post initial gaps near 10^9 and 10^10.** The
optimizer repairs them; the final gaps are 3.1 and 0.002. These are designs that start
absurd and are cheap to fix, which is why `iog` and `fog` disagree about them. Cheap
in the end, but not fast: they need 78 and 17.5 optimizer calls to get within five
percent of the optimum, against 1.5 for the VQGAN, 4 for diffusion and 3.5 for the
lookup table. The unconditional and permuted rows read 101, one more than the budget,
because they never get there. That column is measured against the optimum for each
design's conditions, not against where the design started, so an absurd start does not
read as a fast one.
`gan_cnn_2d` has a `vendi` of 1.09, twelve designs that are effectively one, and a
`cond_sens` of exactly zero, because it is an unconditional model that was handed
conditions anyway. That zero is what stops it collecting a conditional model's score.

**`viol` is high even for real training designs.** On beams2d it includes the volume
budget at a tolerance of 0.01, and a design made for a neighboring condition misses
its own budget by more than that. Seven in twelve of the lookup table's designs fail
on that alone.

Two things this board cannot show. `cond_sens` and the cost columns are blank for the
constructions, because they have no model to re-run or inspect. And there is no
reference row here; that row comes from the saved-designs path in section 7, where at
twelve conditions it is six designs against six, a coarser measurement than the model
rows. The gap path is not always monotone: for the diffusion model it rises after the
first call, as the optimizer restores feasibility before it improves the objective,
which is why its gap after two calls is larger than after one.

## 6. Rank, but only on a column that can be ranked

In [8]:
column = "fog" if EXPENSIVE else "mmd"
board.rank(column)[[column, "train_distance_ratio", "per_condition_distance"]].round(4)

,fog,train_distance_ratio,per_condition_distance
diffusion_2d_cond,-0.9128,7.0439,0.1532
lookup_table,-0.7363,0.0000,0.0569
vqgan,-0.6194,1.7389,0.0833
gan_cnn_2d,0.0019,12.7131,0.4447
unconditional,1.4287,0.0000,0.3993
cgan_cnn_2d,3.0758,14.2969,0.4144
permuted_conditions,4.4569,1.0000,0.4231


In [9]:
try:
    board.rank("train_distance_ratio")
except ValueError as e:
    print(e)

'train_distance_ratio' is a diagnostic: it says whether the other columns mean what they look like. Read it beside them; do not rank on it.


## 7. The same questions in another space

Where a metric is computed is an argument, not a property of the metric. The board
kept every model's sampled designs, so they can be re-scored after projecting onto
the leading principal components of the reference designs. Metrics that need the
actual designs, such as the constraint check and the training-distance column, are
skipped there. The kernel bandwidth is the training designs' median distance in the
projected space, found the same way it was in pixel space.

This is the control that asks whether a learned latent space is needed at all. If PCA
of the data answers the same questions, any rotation of the same width would do. A
learned space registers itself with `register_space` and appears here by name.

In [10]:
pca_board = Board(evaluator.problem, reference=REFERENCE, train=TRAIN)
pca_board.evaluate(board.designs, space="pca", width=8, aggregation="median").round(4)

,mmd@pca,dpp@pca,per_condition_distance@pca,coverage@pca,vendi@pca
cgan_cnn_2d,0.4708,0.0008,7.1693,1.0000,1.2525
gan_cnn_2d,0.7570,0.0000,9.2163,0.5833,1.0015
vqgan,0.0030,0.3758,0.4617,1.0000,6.7750
diffusion_2d_cond,0.0548,0.4440,2.0203,1.0000,6.5482
lookup_table,0.0450,0.1659,0.5618,1.0000,5.8951
unconditional,0.0688,0.2280,8.6422,1.0000,4.8833
permuted_conditions,0.0000,0.2091,10.5238,1.0000,7.4443
reference (split-half),0.1696,0.7890,NaN,1.0000,4.9480


## 8. Registering a metric of your own

A metric is a plain function. The evaluator calls it once per model and hands it one
object, the evaluation context, which holds everything that model's evaluation knows:

| On the context | What it is |
|---|---|
| `gen_designs` | the designs the model produced, one per scored condition, as an array of shape `(n, 50, 100)` here |
| `ref_designs` | the reference optimum for each of those conditions, same shape |
| `conditions` | the scored conditions themselves, one row per design |
| `train_designs` | the training split, flattened, for anything that asks "has the model seen this?" |
| `optimization` | the physics results, one gap path per design; touching it runs the optimizer, so declare `cost="expensive"` |
| `reduce(values)` | collapses one value per design to one number the way the spec asked, median here, mean by default |

The decorator adds the four declarations a reader needs: the name, which family of
question it belongs to, whether it needs the simulator, and which direction is better,
with `None` for a column that is read but never ranked on. The first line of the
docstring is the sentence `explain()` shows. Nothing else has to be wired up.

In [11]:
from engiopt.evaluation.registry import register_metric


@register_metric("density_ratio", family="feasibility", cost="cheap", higher_is_better=None)
def density_ratio(context):
    """Material fraction of each generated design over that of the reference optima. One means the same amount of material."""
    generated = context.gen_designs.reshape(len(context.gen_designs), -1).mean(axis=1)  # one fraction per generated design
    reference = context.ref_designs.reshape(len(context.ref_designs), -1).mean(axis=1)  # one per reference optimum
    return context.reduce(generated) / context.reduce(reference)  # each collapsed the way the spec asked, then compared


Board.from_evaluator(evaluator, {}, designs=board.designs, metrics=["volume_error", "density_ratio"]).explain()

,question,direction,picks,real designs score
volume_error,Gap between each design's material fraction and the fraction its conditions requested. Zero means every design hit its volume target.,lower is better,vqgan,None
density_ratio,Material fraction of each generated design over that of the reference optima. One means the same amount of material.,diagnostic,,None


It appears with its question and direction attached, scored under the same spec as
everything else. Declared diagnostic, because using more material is not better or
worse on its own; read it beside `volume_error`, which says whether each design hit
the fraction its conditions asked for.

## 9. Not in this notebook

- **Latent spaces.** `space="lv"` needs a verified autoencoder pinned by the spec. It
  arrives with the LVAE work and plugs into section 7 by name.
- **Uncertainty.** Every physics column here is a median over fifty designs whose
  per-design values already exist. Error bars from resampling those values, and folds
  of the full test split for the cheap metrics, are the next addition.
- **Seeds.** Every row is training seed one. The Hub holds eleven seeds of each
  family's default configuration; the spread across them is larger than the sampling
  spread for some families, and it belongs on an honest board beside these numbers.